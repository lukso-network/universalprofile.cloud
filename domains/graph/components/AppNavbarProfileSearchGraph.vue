<script setup lang="ts">
import { isAddress, keccak256 } from 'web3-utils'

import type { ProfileSearchQuery } from '@/.nuxt/gql/default'
import type { SearchProfileResult } from '@lukso/web-components'

const INPUT_FOCUS_DELAY = 10 // small delay for focusing input after element render

const { isSearchOpen } = storeToRefs(useAppStore())
const { formatMessage } = useIntl()
const isSearching = ref<boolean>(false)
const searchTerm = ref<string | Address | undefined>()
const hasNoResults = ref<boolean>(false)
const results = ref<SearchProfileResult[]>()

const searchResults = async () => {
  if (!searchTerm.value) {
    return
  }

  isSearching.value = true

  const { profiles }: ProfileSearchQuery = await GqlProfileSearch({
    search: `%${searchTerm.value}%`,
  })

  if (graphLog.enabled) {
    graphLog('profileSearch', profiles)
  }

  if (profiles.length === 0) {
    hasNoResults.value = true
    isSearching.value = false
    return
  }

  hasNoResults.value = false
  results.value = []

  for (const hit of profiles) {
    const metadata = validateLsp3Metadata({
      name: hit.name,
      profileImage: hit.profileImages,
    })
    const profile = await browserProcessMetadata(metadata, keccak256)
    results.value.push({
      name: profile?.name,
      address: hit?.id as Address,
      image: profile.profileImage?.[0]?.url,
    })
  }
  isSearching.value = false
}

const handleSearch = async (event: CustomEvent) => {
  searchTerm.value = event.detail.value
  results.value = undefined

  if (!searchTerm.value) {
    isSearching.value = false
    hasNoResults.value = false
    results.value = undefined
    return
  }

  await searchResults()
}

const handleSelect = (event: CustomEvent) => {
  const selection = event.detail.value
  const { address } = selection
  searchTerm.value = address
  results.value = undefined
  isSearchOpen.value = false

  if (isAddress(address)) {
    navigateTo(profileRoute(address))
    searchTerm.value = undefined
  }
}

watchEffect(() => {
  if (isSearchOpen.value) {
    setTimeout(() => {
      const luksoSearch = document
        .getElementById('mobile-search')
        ?.querySelector('lukso-search') as unknown as HTMLElement
      const inputElement = luksoSearch?.shadowRoot?.querySelector('input')
      inputElement?.focus()
    }, INPUT_FOCUS_DELAY)
  }
})
</script>

<template>
  <lukso-search
    name="profile-search"
    :value="searchTerm"
    :placeholder="formatMessage('profile_search_placeholder')"
    :results="JSON.stringify(results)"
    :is-searching="isSearching ? 'true' : undefined"
    :show-no-results="hasNoResults ? 'true' : undefined"
    :no-results-text="formatMessage('profile_search_no_results')"
    is-full-width
    class="w-full"
    @on-search="handleSearch"
    @on-select="handleSelect"
  ></lukso-search>
</template>
