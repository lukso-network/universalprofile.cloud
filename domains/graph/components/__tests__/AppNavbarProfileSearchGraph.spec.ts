import { renderSuspended } from '@nuxt/test-utils/runtime'
import { fireEvent, screen } from '@testing-library/vue'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import AppNavbarProfileSearchGraph from '../AppNavbarProfileSearchGraph.vue'

describe('AppNavbarProfileSearchGraph', () => {
  beforeEach(() => {
    useIntl().setupIntl(defaultConfig)
  })

  it('should render', async () => {
    const component = await renderSuspended(AppNavbarProfileSearchGraph)

    expect(component.html()).toMatchSnapshot()
  })
})
