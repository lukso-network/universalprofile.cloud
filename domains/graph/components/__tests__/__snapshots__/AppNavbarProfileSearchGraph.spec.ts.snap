// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppNavbarProfileSearchGraph > should render 1`] = `
"<div id="test-wrapper">
  <lukso-search name="profile-search" placeholder="Search profiles" no-results-text="No profile found" is-full-width="" class="w-full"></lukso-search>
</div>"
`;
